File: imageSlicer.py
import os
import sys
from PIL import Image

# raised when num_slices is not a divisor of instagram_wh
class InstagramSizeException(Exception):
  pass

class ImageSlicer:
  """
  Image slicer

  Attributes:
    imgsrc     -- string specifiying the location of the image to be sliced
    num_slices -- number of slices to cut the image into. Ex: 12 = (12 x 12)
    saveloc    -- location to save the resulting slices 
  """

  # default width and height of instagram photos.
  instagram_wh = 612

  def __init__(self, imgsrc, num_slices=12, saveloc="pieces/"):
    self.imgsrc = imgsrc
    self.num_slices = num_slices
    self.saveloc = saveloc
    self.slices = []
    self.rgbs = []
    if ImageSlicer.instagram_wh % num_slices != 0:
      raise InstagramSizeException("num_slices is not a divisor of 612")

  # slice image into pieces of equal size defined by num_slices
  def slice(self):
    if (self.slices != []):
      return self.slices
    img = Image.open(self.imgsrc)
    width, height = img.size
    slices = [[None for w in range(self.num_slices)] for h in range(self.num_slices)]
    if (width != ImageSlicer.instagram_wh or 
       height != ImageSlicer.instagram_wh):
      raise InstagramSizeException("Photo is not 612 x 612")

    subwidth = ImageSlicer.instagram_wh / self.num_slices
    subheight = ImageSlicer.instagram_wh / self.num_slices
    for x in xrange(self.num_slices):
      for y in xrange(self.num_slices):
        box = [
          x * subwidth,
          y * subheight,
          (x + 1) * subwidth,
          (y + 1) * subheight
        ]
        piece = img.crop(box)
        slices[y][x] = (piece)
    self.slices = slices
    return slices

  def get_RGB_values(self):
    if self.slices is None or self.slices == []:
      self.slice()
    self.rgbs = map(lambda y: map(lambda x: list(x.getdata()), y), self.slices)
    return self.rgbs
    
  def average_RGB(self):

    cut_size = 3
    # def reduce_average((r1,g1,b1), (r2,g2,b2)):
    #   return ((r1 + r2) / 2, (g1 + g2) / 2, (b1 + b2) / 2)

    def reduce_average(l):
      rf = 0
      gf = 0
      bf = 0
      for (r,g,b) in l:
        rf += r
        gf += g
        bf += b
      num = len(l)
      return (rf / num, gf / num, bf / num)

    subpieces = []
    if (self.rgbs == []):
      self.get_RGB_values()
    subwidth = ImageSlicer.instagram_wh / self.num_slices
    subheight = ImageSlicer.instagram_wh / self.num_slices

    average_subwidth = subwidth / cut_size
    average_subheight = subheight / cut_size

    averages = []
    for x in xrange(self.num_slices):
      for y in xrange(self.num_slices):
        current_slice = self.slices[x][y]
        for i in xrange(cut_size):
          for j in xrange(cut_size):
            sub_averages = []
            cropped = current_slice.crop((average_subwidth * j,
                                          average_subheight * i,
                                          average_subwidth * j + average_subwidth,
                                          average_subheight * i + average_subheight))
            sub_averages.append(reduce_average(list(cropped.getdata())))
            averages.append(sub_averages)
    
    flattened = []
    block_size = self.num_slices*cut_size**2
    h = 0
    while (h != self.num_slices):
      for k in range(block_size*h, block_size*(h+1)):
        if (k % (cut_size**2) == 0 or k % (cut_size**2) == 1 or k % (cut_size**2) == 2):
          flattened.append(averages[k])
      
      for m in range(block_size*h, block_size*(h+1)):
        if (m % (cut_size**2) == 3 or m % (cut_size**2) == 4 or m % (cut_size**2) == 5) :
          flattened.append(averages[m])
      
      for n in range(block_size*h, block_size*(h+1)):
        if (n % (cut_size**2) == 6 or n % (cut_size**2) == 7 or n % (cut_size**2) == 8) :
          flattened.append(averages[n])          
      h+=1
    
    return [item for sublist in flattened for item in sublist]


slicer = ImageSlicer("images/440815042347915310_25264853.jpg", 51)

# slicer.average_RGB()
# exit(0)

averages = slicer.average_RGB()

img = Image.new('RGB', (153, 153))
# for x in range(36):
#   for y in range(36):
#     img_data[y,x] = averages[(x * 36) + y]

img.putdata(averages)
img.save("./test3.jpg")
# # Image manip
# img = Image.open(path)
# # get a list of all rgb values in the image
# list(img.getdata())

File: imageSlicer.cpp
/*
ImageSlicer.cpp designed to reproduce the results of
ImageSlicer.py and slice up a target image into sections
used in image matching for creating photo mosaics
*/

#include <iostream>
#include <string>
#include <vector>
#include <string>
#include <sstream>
#include <Magick++.h>
#include <math.h>
#include <climits>
#include <list>
#include "mongo/client/dbclient.h"
#include "CycleTimer.h"

using namespace std;
using namespace Magick;
using namespace mongo;

struct RGB {
  int red;
  int green;
  int blue;
};

class ImageSlicer {
  string imgsrc; // source of the input image
  int numSlices; // number of slices in x and y direction
  int cutSize; // number of sub slices to make from each image 
  vector< vector< RGB > > rgbs; // vector to store RGB values of pieces
  vector<RGB> averages; // vector to store average rgbs
  vector< vector< Image > > slices; // image slices.
  Image sourceImage;

public:
  ImageSlicer(string imgsrc, int numSlices);
  vector<RGB> getAverages();
  vector< vector< Image > > getSlices();

private:
  void slice();
  void calculateRGBValues();
};

ImageSlicer::ImageSlicer (string src, int n) {
  imgsrc = src;
  numSlices = n;
  cutSize = 3;
  slices = vector< vector<Image> > (51);
  for (int x = 0; x < n; x++) {
    slices[x] = vector<Image> (51);
  }
  sourceImage = Image(src);
}

void ImageSlicer::slice() {
  Image img = sourceImage;
  int width = img.columns();
  int height = img.rows(); // size of instagram photo

  int subwidth = width / numSlices;
  int subheight = height / numSlices;
  int count = 0;
  for (int x = 0; x < numSlices; x++) {
    for (int y = 0; y < numSlices; y++) {
      Image piece = img;
      piece.crop(Geometry(subwidth, subheight, x * subwidth, y * subheight));
      // piece.write(sstr.str());
      slices[y][x] = piece;
      count++;
    }
  }
  return;
}

void ImageSlicer::calculateRGBValues() {
  int subwidth = 612 / numSlices;
  int subheight = 612 / numSlices;

  int average_subwidth = subwidth / cutSize;
  int average_subheight = subheight / cutSize;

  for (int x = 0; x < numSlices; x++) {
    for (int y = 0; y < numSlices; y++) {
      Image currentSlice = slices[x][y];
      Pixels view(currentSlice);
      PixelPacket *pixels = view.get(0, 0, currentSlice.columns(), currentSlice.rows());
      for (int i = 0; i < cutSize; i++) {
        for (int j = 0; j < cutSize; j++) {
          int red = 0;
          int green = 0;
          int blue = 0;
          for (int k = 0; k < average_subwidth; k++) {
            for (int l = 0; l < average_subheight; l++) {
              int curX = (i * average_subwidth) + k;
              int curY = (j * average_subheight) + l;
              int pixelLoc = curX + (curY * subwidth);
              ColorRGB pixel = ColorRGB(pixels[pixelLoc]);
              red += pixel.red() * 255;
              green += pixel.green() * 255;
              blue += pixel.blue() * 255;
            }
          }
          int numIters = average_subwidth * average_subheight;
          red /= numIters;
          blue /= numIters;
          green /= numIters;
          RGB rgb;
          rgb.red = red;
          rgb.green = green;
          rgb.blue = blue;
          averages.push_back(rgb);
        }
      }
    }
  }
}

vector< vector< Image > > ImageSlicer::getSlices() {
  slice();
  return slices;
}

vector<RGB> ImageSlicer::getAverages() {
  slice();
  calculateRGBValues();
  return averages;
}

int square(int x) {
  return x * x;
}

int RGBdistance(RGB t1, RGB t2) {
  int red = square(t1.red - t2.red);
  int green = square(t1.green - t2.green);
  int blue = square(t1.blue - t2.blue);
  return (int)sqrt(red + green + blue);
}

int totalDistance(vector<RGB> a1, vector<RGB> a2) {
  if (a2.size() != 9) {
    return 255 * 255;
  }
  int dist = 0;
  for (size_t i = 0; i < a1.size(); i++) {
    dist += RGBdistance(a1[i], a2[i]);
  }
  return dist;
}

int main(int argc, char **argv) {
  InitializeMagick(*argv);
  if (argc < 3) {
    cout << "usage: " << argv[0] << " <image path> <save path>\n";
    return 1;
  }
  ImageSlicer slicer(argv[1], 51);
  string savepath = argv[2];
  DBClientConnection c;
  c.connect("localhost");
  vector <vector <RGB> > dbImageColors;
  vector <string> dbImageSources;
  auto_ptr<DBClientCursor> cursor = c.query("instagram_photomosaic.image_pool", BSONObj());
  // load all the stuff from the database to check against.
  double dbstart = CycleTimer::currentSeconds();
  while (cursor->more()) {
    BSONObj obj = cursor->next();
    dbImageSources.push_back(obj.getStringField("imgsrc"));
    BSONObjIterator fields (obj.getObjectField("averages"));
    vector <RGB> curRGBs;
    while (fields.more()) {
      vector<BSONElement> elems = fields.next().Array();
      int red = elems[0].Int();
      int green = elems[1].Int();
      int blue = elems[2].Int();
      RGB rgb;
      rgb.red = red;
      rgb.green = green;
      rgb.blue = blue;
      curRGBs.push_back(rgb);
    }
    // add vector of 9 rgbs to large vector
    dbImageColors.push_back(curRGBs);
  }
  double dbend = CycleTimer::currentSeconds();
  printf("Time to read in DB %f\n", (dbend - dbstart));
  // average values of the input image. This is an array of a bunch of RGBs
  // where they are grouped in 9s in order.
  double avgstart = CycleTimer::currentSeconds();
  vector<RGB> averages = slicer.getAverages();
  double avgend = CycleTimer::currentSeconds();
  printf("Time to find averages of input image %f\n", (avgend - avgstart));
  vector<string> finalImages;
  double imgstart = CycleTimer::currentSeconds();
  for (size_t i = 0; i < averages.size(); i += 9) {
    // current value of the minimum distance and it's index.
    int minIndex = 0;
    int minVal = INT_MAX;
    vector<RGB> current;
    // grab the next 9 values (corresponds to a subimage)
    for (int j = 0; j < 9; j++) {
      current.push_back(averages[i + j]);
    }
    for (size_t k = 0; k < dbImageColors.size(); k++) {
      // get distance from current subimage to current image from db.
      int dist = totalDistance(current, dbImageColors[k]);
      if (dist < minVal) {
        minIndex = k;
        minVal = dist;
      }
    }
    finalImages.push_back(dbImageSources[minIndex]);
    // replace minIndex with empty vector to remove values and avoid duplicates.
    vector<RGB> empty;
    dbImageColors[minIndex] = empty;
  }
  double imgend = CycleTimer::currentSeconds();
  printf("Time to compute image matches %f\n", (imgend - imgstart));
  list <Image> finalMontage;
  Montage montage;
  montage.tile("51x51");
  montage.geometry("48x48");
  vector<Image> images;
  double montagestart = CycleTimer::currentSeconds();
  for (size_t n = 0; n < finalImages.size(); n++) {
    string filename = finalImages[n];
    Image mosaicImage(filename);
    mosaicImage.resize("48x48");
    if (n % 50 == 0) {
      printf("Opening image %zu\n", n);
    }
    images.push_back(mosaicImage);
  }
  printf("Attempting to montage the image\n");
  montageImages(&finalMontage, images.begin(), images.end(), montage);
  writeImages(finalMontage.begin(), finalMontage.end(), savepath);
  double montageend = CycleTimer::currentSeconds();
  printf("Time to create and write montage to file %f\n", (montageend - montagestart));
}



